### DISCUSSION TITLE
Opposite of `trailing_comma_in_multiline`?

### DISCUSSION THREAD

**Discussion by blueblakk** [PHP-CS-Fixer/PHP-CS-Fixer#6814] [Q&A] [answered]
I'm trying to not have trailing comma's in multiline, but when I configure it the following way it doesn't remove them:

`
'trailing_comma_in_multiline' => ['elements' => []]
`

Does it not support this?

**Comment by kubawerlos** [comment]
Currently, fixer `trailing_comma_in_multiline` is for adding commas as configured and you have configured it to add it nowhere.

Is not having trailing commas a part of some community coding standards? If so the adding flag `strict` (similar as in other fixers) could be an enhancement.

**Reply by keradus** [reply, marked as answer]
I assume there is none, never heard of. If there is and larger community would like to have it, we can consider to bring it up. For now, closing this as not enough support for the idea.

**Comment by zerocrates** [comment]
The use I can see for this would be kind of specific to parameters, to enforce 7x compatibility by banning parameter trailing commas.

**Comment by mlocati** [comment]
For example, in PHP CS Fixer itself in the `DescribeCommand.php` we currently have:

- [at lines 276-279](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.92.3/src/Console/Command/DescribeCommand.php#L276-L279):
  ```php
  $allowed = array_map(
      static fn (string $type): string => '<comment>'.$type.'</comment>',
      $allowedTypes,
  );
  ```
- [at lines 433-437](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.92.3/src/Console/Command/DescribeCommand.php#L433-L437):
  ```php
  $rulesWithoutSet = array_filter(
      $this->getFixers(),
      static fn (string $name): bool => [] === FixerDocumentGenerator::getSetsOfRule($name),
      \ARRAY_FILTER_USE_KEY
  );
  ```

At the moment there's no way to force "no trailing commas in call arguments".

A more flexible configuration of the `trailing_comma_in_multiline` fixer, instead of an `elements` array containing the elements that must have trailing commas, could be something like
```php
[
    'arguments' => bool|null,
    'array_destructuring' => bool|null,
    'arrays' => bool|null,
    'match' => bool|null,
    'parameters' => bool|null,
]
```
Where `bool|null` can be `true` (force yes), `false` (force no), `null` (don't touch existing code).

**Reply by keradus** [reply]
Part of Fixer philosophy is not to add option to configure everything that is possible, but only to those we consider best practices or widely adopted standard.

the diff between existing option and suggested option would be possibility to _remove_ the comma. is there any standard expecting this, other than "let's aim to support legacy PHP" ?

**Reply by keradus** [reply]
> (..) we currently have (...)

#9314

**Comment by mlocati** [comment]
One never knows all the possible use cases of tools...

For example, let's assume that:

- I'm developing a library with 2 versions: "edge" and "legacy"
- for the "edge" version I have `'trailing_comma_in_multiline' => ['elements' => ['arrays', 'parameters']]`
- for the "legacy" version I have `'trailing_comma_in_multiline' => ['elements' => ['arrays']]`
- I push a commit to "edge"

It'd be nice to have PHP CS Fixer fix that commit automatically for "legacy", by removing commas in multiline method definitions.
